pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -898,61 +898,43 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, StablePoolStorage,
-    {
-        uint256[] memory balances = new uint256[](balancesWithoutBpt.length);
-        uint256[] memory ratiosWithoutBpt = _dropBptItem(_getTokenRateRatios());
-        for (uint256 i = 0; i < balancesWithoutBpt.length; ++i) {
-            balances[i] = balancesWithoutBpt[i].mulDown(ratiosWithoutBpt[i]);
-        }
-
-        return balances;
-    }
-
-    /**
-     * @dev Return the complete set of token ratios (including BPT, which will always be 1).
-     */
-    function _getTokenRateRatios() internal view returns (uint256[] memory rateRatios) {
-        uint256 totalTokens = _getTotalTokens();
-        rateRatios = new uint256[](totalTokens);
-
-        // The Pool will always have at least 3 tokens so we always load these three ratios.
-        rateRatios[0] = _isTokenExemptFromYieldProtocolFee(0)
-            ? _computeRateRatio(_tokenRateCaches[_getToken0()])
-            : FixedPoint.ONE;
-        rateRatios[1] = _isTokenExemptFromYieldProtocolFee(1)
-            ? _computeRateRatio(_tokenRateCaches[_getToken1()])
-            : FixedPoint.ONE;
-        rateRatios[2] = _isTokenExemptFromYieldProtocolFee(2)
-            ? _computeRateRatio(_tokenRateCaches[_getToken2()])
-            : FixedPoint.ONE;
-
-        // Before we load the remaining ratios we must check that the Pool contains enough tokens.
-        if (totalTokens == 3) return rateRatios;
-        rateRatios[3] = _isTokenExemptFromYieldProtocolFee(3)
-            ? _computeRateRatio(_tokenRateCaches[_getToken3()])
-            : FixedPoint.ONE;
-
-        if (totalTokens == 4) return rateRatios;
-        rateRatios[4] = _isTokenExemptFromYieldProtocolFee(4)
-            ? _computeRateRatio(_tokenRateCaches[_getToken4()])
-            : FixedPoint.ONE;
-
-        if (totalTokens == 5) return rateRatios;
-        rateRatios[5] = _isTokenExemptFromYieldProtocolFee(5)
-            ? _computeRateRatio(_tokenRateCaches[_getToken5()])
-            : FixedPoint.ONE;
+    function _getAdjustedBalances(uint256[] memory balances) internal view returns (uint256[] memory adjustedBalances) {
+        uint256 totalTokens = balances.length;
+        adjustedBalances = new uint256[](totalTokens);
+
+        // The Pool will always have at least 3 tokens so we always adjust these three balances.
+        adjustedBalances[0] = _isTokenExemptFromYieldProtocolFee(0)
+            ? _adjustedBalance(balances[0], _tokenRateCaches[_getToken0()])
+            : balances[0];
+        adjustedBalances[1] = _isTokenExemptFromYieldProtocolFee(1)
+            ? _adjustedBalance(balances[1], _tokenRateCaches[_getToken1()])
+            : balances[1];
+        adjustedBalances[2] = _isTokenExemptFromYieldProtocolFee(2)
+            ? _adjustedBalance(balances[2], _tokenRateCaches[_getToken2()])
+            : balances[2];
+
+        // Before we adjust the remaining balances we must check that the Pool contains enough tokens.
+        if (totalTokens == 3) return adjustedBalances;
```

Comment:
I'd also add an explicit return at the end for consistency. Also drop the named return argument if we're not using it.